src/XGridCtrl.cpp

```diff
@@ -1665,7 +1665,7 @@ XGridCtrl::OnArrow(puz::GridDirection arrowDirection, int mod)
                  newSquare;
                  newSquare = newSquare->Next(arrowDirection))
             {
-                if (newSquare->HasWord(focusedDirection)
+                if ((newSquare->HasWord(focusedDirection) || m_puz->FindWord(newSquare) != NULL)
```

Comment:
I think you need `FindWord(newSquare, focusedDirection)`, otherwise this is also going to find words in the crossing direction.

IIRC this logic is for things like cryptics, so that you don't have to run through all the single squares that aren't actually part of a word.

Reply:
Good point, done.

For cryptics - every single square is part of a word; just potentially in a different direction. FWIW, in that case, shift arrow doesn't quite work like I'd expect for the other case that the arrow is in the opposite direction of the current one (so independent of my change). If you're in the across direction in a square that crosses a clue in the down direction, and you shift-down, you move down one square and shift directions down:

https://github.com/mrichards42/xword/blob/master/src/XGridCtrl.cpp#L1680-L1686

I'd expect it to move to the next white square that's part of a word in the current direction - otherwise, it's just doing the same thing as the arrow keys, no?

Reply:
Yeah, good point. Maybe the logic for both cases is just this:

```
for square over Next(arrowDirection):
  if a new word exists in focusedDirection:
    break
```

... except for arrowDirection == focusedDirection you'd wrap around the grid, and otherwise you wouldn't wrap.

Reply:
Makes sense to me. Pushed a new commit which implements that.

One slight weird case is for Acrostics - if you navigate into the quote, you can no longer shift-navigate out, because the quote has a unique direction (since it's multiple lines long) that won't match any of the regular "across" clues below the quote. That being said, this feels like a bit of an edge case to me, and it probably makes more sense to just navigate with arrow keys / enter+shift-enter in that case?

Reply:
Not to complicate things further, but should acrostics just translate shift+arrow -> arrow? Or is it something you'll just figure out on your own if you're solving an acrostic?

Reply:
There may be other cases as well - I tried a Rows Garden and it also doesn't navigate well between bloom clues (which are multiple rows).

Reply:
Digging into this a bit deeper:

- Rows Garden - the issue is that once we find a new square, we call newSquare->GetWordStart(focusedDirection) to move to the start of the word containing that square. For reasons I don't quite understand, GetWordStart then inverts the provided direction and finds the boundary in the opposite direction, which doesn't exist for Rows Gardens. So every Shift+Arrow navigation with a bloom selected just shoots you to the top (since that would be the theoretical start of a word in the "Down" direction). I don't really understand why we're inverting the direction to find the start of a word, but one simple fix here is to use the actual Word we found (if we found one) and take the first square of that instead, and only use GetWordStart if we found a "possible" word.

- Acrostics - the issue here is indeed that the quote word has a weird direction, since it calculates it from the first square in the top left of the quote to the last square in the bottom right. So there's no next word with the same direction, since it's the only word in the grid with that direction.

  The best I can come up with is - if we can't find a word in the exact same direction as calculated by this math, we can instead look for the first word we can find with a clue that's in the same clue list as the current word (if any). Since Acrostics have only one clue list, this seems like it should work. It still doesn't feel perfect to me - it seems like you could plausibly define an Acrostic where the grid quote is a separate clue list from the individual clues - but it's probably the most reasonable thing we could do here without making navigation weirder for other puzzles? If that makes sense to you - what's the most efficient way to go from a square to "word containing this square with a clue in the same clue list in the current clue"?

  I tried your proposal of translating shift+arrow->arrow for Acrostics; it mostly makes sense, and I'd probably be fine with that, but it does slow things down a bit when moving between the two columns of clues below the quote.

Reply:
> For reasons I don't quite understand, GetWordStart then inverts the provided direction and finds the boundary in the opposite direction ...

Reading through that code I think it's just that the ACROSS direction means "to the right", so to get to the start of the across word you invert it to go "to the left" (invert meaning 180, not ACROSS <-> DOWN).

> So every Shift+Arrow navigation with a bloom selected just shoots you to the top (since that would be the theoretical start of a word in the "Down" direction). [...] one simple fix here is to use the actual Word we found (if we found one) and take the first square of that instead, and only use GetWordStart if we found a "possible" word.

This logic makes sense to me. I _think_ what's going on is that focusedDirection is something like -33 degrees, which means [no word start](https://github.com/mrichards42/xword/blob/1461066057b31638c6899a24e28c75c222b6a48c/puz/Square.cpp#L186-L189).

FWIW since this started as a puz app, words used to go only straight across or down and could only be white squares . . . iirc puz::Word was added later for jpz support, so I wouldn't be surprised if there is more code around that uses the puz definition of a word.

I'm pretty sure real Words exist for all puzzles at this point, the only exception being that diagramless puzzles always use "invented" words (which looks like it happens in [XGridCtrl::DoSetFocusedWord](https://github.com/mrichards42/xword/blob/f6bf00d9b21f043e17c48daf9d200a03ad348a19/src/XGridCtrl.cpp#L880-L903)).


> Acrostics - the issue here is indeed that the quote word has a weird direction, since it calculates it from the first square in the top left of the quote to the last square in the bottom right.

As in it gets some arbitrary angle? I guess that's more leftovers from assuming words go in a straight line.

> if we can't find a word in the exact same direction as calculated by this math, we can instead look for the first word we can find with a clue that's in the same clue list as the current word (if any). [...] what's the most efficient way to go from a square to "word containing this square with a clue in the same clue list in the current clue"?

That works for me. I think this is the same logic as Tab (or shift+Tab for the up or left arrow), so probably take a look there for the cluelist part.

Reply:
Alright - uploaded a commit which implements the "first word we can find with a clue in the same clue list". I think it's different from Tab/Shift+Tab in that the clue will not always be the next one in the list - instead it's whatever clue contains the closest square in the arrow direction. Seems to work as I'd expect for Acrostics, and hopefully doesn't come up too often if ever for other puzzle types with relatively standard directions.